glib/src/value.rs

```diff
@@ -56,31 +51,262 @@ use crate::gstring::GString;
 use crate::translate::*;
 use crate::types::{StaticType, Type};
 
-/// An error returned from the [`get`](struct.Value.html#method.get)
-/// or [`get_some`](struct.Value.html#method.get_some) functions on a [`Value`](struct.Value.html)
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct GetError {
-    pub actual: Type,
-    pub requested: Type,
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling optional types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is optional.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeOptional:
+    ValueType
+    + ToValueOptional
+    + for<'r> FromValue<'r, Error = WrongValueTypeOrNoneError<Self>>
+    + StaticType
+{
+}
+
+impl<'r, T: ValueTypeOptional + StaticType> ValueType for Option<T>
+where
+    Self: for<'s> FromValue<'s, Error = WrongValueTypeOrNoneError<T>>,
+{
+    type Type = T::Type;
+}
+
+/// Trait for errors returned from `Value::get`.
+pub trait GetValueError: std::error::Error + Sized {
```

Comment:
This trait should probably be `unsafe` because you can probably cause memory unsafety if implemented by simply returning `Ok(())` :)

Reply:
Hum, I guess this is also the case for `TryFromGlib`.

Reply:
Can you create an issue for that so we don't forget?

Reply:
https://github.com/gtk-rs/gtk-rs/issues/462